### DISCUSSION TITLE
How can I emit and receive data using strapi-plugin-io?

### DISCUSSION THREAD

**Discussion by japineda16** [strapi-community/plugin-io#58] [General]
I achieve connect my front end (made in nextJS), and my backend, but when I want to use the function io.emit(), and other side use  io.on(), using same channel, it doesn’t work. Somebody has any experience using that plugin in Strapi?

**Comment by ComfortablyCoding** [comment]
Can you show what the relevant frontend and backend code looks like?

**Reply by japineda16** [reply]
I achieve connect frontend and backend, I create two events in config file, but now I'd like to create a channel for each user, can you show me an example to how to send and read some date from a exclusive channel?

**Reply by japineda16** [reply]
This is my socket.io backend code
```
module.exports = ({ env }) => ({
  // ...
  io: {
    enabled: true,
    config: {
      IOServerOptions: {
        cors: { origin: "http://localhost:3000", methods: ["GET"] },
      },
      contentTypes: {
        message: "*",
        chat: ["create"],
        vehicles: "*",
      },
      events: [
        {
          name: "connection",
          handler: ({ strapi }, socket) => {
            strapi.log.info(`[io] new connection with id ${socket.id}`);
            socket.on("setLocation", (userId) => {
              socket.join(userId);
              setTimeout(() => {
                socket.emit("getLocation", userId);
              }, 500);
            });
          },
        },
      ],
    },
  },
  // ...
});
```

**Reply by japineda16** [reply]
```
  useEffect(() => {
    socket.on("connect", () => {
      socket.emit("setLocation", user.id);
      socket.on("getLocation", (data) => console.log(data));
    });
  }, []);
```
Here is my front end code, I want to send user id as channel name, where my front end will receive all data from a mobile app, so the idea is subscribe frontend web to strapiIO (backend), which this one will receive some coords, and mobile app will send those coords.

**Reply by ComfortablyCoding** [reply]
I cannot replicate this issue, for me both `setLocation` and `getLocation` trigger at the expected times.

I used a regular html file over nextjs, so it is possible something else you are doing is causing this.

**Reply by japineda16** [reply]
Update: I can connect to both channels, and it is working fine, thank you for the help, but now I'd like to know how can I create a private room? Can you show me an example? I want to create a private room per user, where I'm gonna send the coords.

**Reply by ComfortablyCoding** [reply]
I believe every socket automatically has their own room by default. The room name is the socket id. 

Alternatively when they join you can add them to a room. By ensuring its unique per socket will make it "private".

I don't have any specific example code, this is more a general socket io question than anything specific to this plugin.

**Reply by japineda16** [reply]
Hi, it working fine, but only with massive messages, I'd like to know just how to make private messages, I tried using method join (as I used to do in Socket.IO), but it doesn't work when I want to send private messages, only massive, let me show you my code
**Strapi:**
```
          handler: ({ strapi }, socket) => {
            console.log("a user connected");
            socket.on("location", ({ userId, latitude, longitude, speed }) => {
              console.log("location", userId, latitude, longitude, speed);
              // making a room
              socket.join(userId);
              // sending message to this room.
              socket
                .to(userId)
                .emit("location", { latitude, longitude, speed });
            });
          },
```
Front end NextJS:
```
  useEffect(() => {
    socket.emit("location", {
      latitude: 0,
      longitude: 0,
      userId: useAuthStore.getState().user.id ?? null,
    });
    socket.on("location", (data) => {
      console.log(data);
      socket.on({userId}, (data) => {
        console.log(data);
      });
    });
  }, []);
```

**Reply by ComfortablyCoding** [reply]
On the server side you need to be using the server socket to send an event to a client side socket. The socket you are using right now is the client socket.

Your server side code (in this case) should look something along the lines of 
```js
handler: ({ strapi }, socket) => {
  console.log("a user connected");
  socket.on("location", ({ userId, latitude, longitude, speed }) => {
    console.log("location", userId, latitude, longitude, speed);
    // making a room
    socket.join(userId);
    // sending message to this room.
    strapi.$io.raw(
      "location",
      { latitude, longitude, speed },
      { room: userId }
    );
  });
},
```

This should trigger the `socket.on("location",....);` you have on your client side.
